aten/src/ATen/native/cuda/ReduceSumProdKernel.cu

```diff
@@ -36,8 +37,9 @@ static void sum_kernel_cuda(TensorIterator& iter) {
     return sum_kernel_impl<at::BFloat16, float, float>(iter);
   }
   #endif
-  AT_DISPATCH_ALL_TYPES_AND(ScalarType::Bool, iter.dtype(), "sum_cuda", [&]() {
-    sum_kernel_impl<scalar_t>(iter);
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(ScalarType::Bool, iter.dtype(), "sum_cuda", [&]() {
+    using value_t = typename ztype<scalar_t>::value_t;
```

Comment:
I guess `AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX_AND` and remove the `ztype` will just work.

Reply:
hmm there was an issue with `__shfl_up_sync` for c10::complex. I'll look into it more

Reply:
use `::thrust_t`